### PR TITLE
open dygraph op test, test=develop

### PR DESCRIPTION
**Pull request by JepsonWong**
open dygraph op test.
**The PR solves the problem of opening op test in dygraph.**

**There are several types of problems:**
**1.Does not support lod in dygraph.** 

Through the Variable._ivar.value().get_tensor().set_recursive_sequence_lengths interface, I can set the lod information for the LoDTensor, so that the following ops that do not depend on the scope structure pass the test, but the interface is not exposed to the user. Debug the lod test is only for verifying the correctness of the dygraph optest framework.

**Some lod information needs to be read from the scope structure.**

- test_attention_lstm_op.py (成功)
- test_fused_emb_seq_pool_op.py
- test_fused_embedding_fc_lstm_op.py (成功)
- test_fusion_gru_op.py (成功)
- test_fusion_lstm_op.py (成功)
- test_fusion_seqconv_eltadd_relu_op.py (成功)
- test_lod_reset_op.py
- test_lstm_op.py (成功)
- test_multiclass_nms_op.py (成功)
- test_one_hot_op.py (成功)
- test_one_hot_v2_op.py (成功)
- test_seq_pool.py (成功)
- test_sequence_pad_op.py (成功)
- test_warpctc_op.py (成功)
- linear_chain_crf_op (成功)

**2.Does not support mkldnn in dygraph.**

- test_activation_mkldnn_op.py
- test_concat_int8_mkldnn_op.py
- test_concat_mkldnn_op.py
- test_conv2d_int8_mkldnn_op.py
- test_dequantize_mkldnn_op.py
- test_fc_mkldnn_op.py
- test_lrn_mkldnn_op.py
- test_mul_int8_mkldnn_op.py
- test_pool2d_int8_mkldnn_op.py
- test_quantize_mkldnn_op.py
- test_requantize_mkldnn_op.py
- test_softmax_mkldnn_op.py
- test_sum_mkldnn_op.py
- test_transpose_int8_mkldnn_op.py
- test_transpose_mkldnn_op.py

The following optest cannot be passed when uses mkldnn:

- test_conv2d_op.py
- test_conv2d_transpose_op.py
- test_conv3d_op.py
- test_elementwise_add_op.py
- test_elementwise_mul_op.py
- test_pool2d_op.py
- test_softmax_op.py

**3.Op constructs the reverse OpDesc using the DefaultGradOpDescMaker class, but some output grad is not used.**

- nce_op
- unpool_op
- var_conv_2d_op

**4.The HasInput and HasOutput methods in ExecutionContext depend on the Scope structure, while the Scope structure in the dynamic graph is empty.**

- strided_slice_op
- activation_op
- crf_decoding_op
- crop_op
- crop_tensor_op
- data_norm_op
- edit_distance_op
- expand_op
- fused_elemwise_activation_op
- im2sequence_op
- reshape_op
- sequence_mask_op
- slice_op

**5. Execution function depend on the Scope structure**

- test_lstm_cudnn_op.py

**6.In dygraph RuntimeInferVarTypeContext class, if you want to determine the existence of a variable, you should use the HasOutput method instead of the Output method.**

- hierarchical_sigmoid_op

**7.Exception capture in optest.**

- test_fill_any_like_op.py

**8.Reverse OpDesc class is not standardized.**

- test_increment_ngraph_op

**At present, because dygraph does not support lod and mkldnn, the ops in problems 1, 2 have been banned, and problem 5, 7, 8 have been banned.
Problems 3, 4, and 6 have been resolved.**

CI耗时统计：
MAC_PR_CI: 36m51s
MAC_PR_CI_python35: 38m16s
PR_CI_Coverage: 1h23m
PR_CI_GPU_Py2_GCC8: 44m59s
PR_CI_Inference: 15m51s
PR_CI_python35: 40m10s
PR_Windows_CI: 1h32m

### CODE REVIEW

File: paddle/fluid/operators/crf_decoding_op.h [hunk 1]
```diff
@@ -39,7 +39,7 @@ class CRFDecodingOpKernel : public framework::OpKernel<T> {
     math::SetConstant<DeviceContext, int64_t>()(
         ctx.template device_context<DeviceContext>(), decoded_path, 0);
 
-    bool has_length = ctx.HasInput("Length");
+    bool has_length = ctx.InputVar("Length") != nullptr;
```
Comment (JiabinYang) [inline paddle/fluid/operators/crf_decoding_op.h]: Why not change `HasInput`

File: paddle/fluid/operators/crop_op.h [hunk 2]
```diff
@@ -30,10 +30,11 @@ using framework::Tensor;
 static std::vector<int> GetOffsets(const framework::ExecutionContext& ctx) {
   std::vector<int> res;
   int rank = ctx.Input<Tensor>("X")->dims().size();
-  if (ctx.HasInput("Offsets")) {
-    PADDLE_ENFORCE(ctx.Attr<std::vector<int>>("offsets").empty(),
-                   "Input 'Offsets' and attribute 'offsets' should not be used "
-                   "at the same time.");
+  if (ctx.InputVar("Offsets") != nullptr) {
```
Comment (JiabinYang) [inline paddle/fluid/operators/crop_op.h]: same above

File: paddle/fluid/operators/crop_tensor_op.h [hunk 3]
```diff
@@ -92,7 +92,7 @@ static std::vector<int> GetShape(const framework::ExecutionContext& ctx) {
     return res;
   }
 
-  auto* shape_tensor = ctx.HasInput("Shape")
+  auto* shape_tensor = ctx.InputVar("Shape") != nullptr
```
Comment (JiabinYang) [inline paddle/fluid/operators/crop_tensor_op.h]: same and all below

File: paddle/fluid/operators/data_norm_op.cc [hunk 4]
```diff
@@ -309,7 +309,7 @@ class DataNormGradKernel<platform::CPUDeviceContext, T>
 
     // init output
     Tensor *d_x = nullptr;
-    if (ctx.HasOutput(framework::GradVarName("X"))) {
+    if (ctx.OutputVar(framework::GradVarName("X")) != nullptr) {
```
Comment (JiabinYang) [inline paddle/fluid/operators/data_norm_op.cc]: same as `HasInput`

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 5]
```diff
@@ -277,45 +278,142 @@ def _calc_dygraph_output(self, place, parallel=False, no_check_set=None):
...
-            for name, np_value in six.iteritems(self.inputs):
-                if not isinstance(np_value, list):
-                    np_value = [np_value]
-
-                for i in range(len(np_value)):
-                    inputs[name].append(
-                        self._create_var_from_numpy(np_value[i]))
-
+            op_proto = OpProtoHolder.instance().get_op_proto(self.op_type)
+            for input_ in op_proto.inputs:
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/op_test.py]: seal these code as a `method`

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 6]
```diff
@@ -963,11 +1079,248 @@ def check_grad_with_place(self,
...
+                        return output_vars_selected
+
+    def _get_dygraph_grad(self,
+                          inputs_to_check,
+                          place,
+                          output_names,
+                          no_grad_set=None):
+        with fluid.dygraph.base.guard(place=place):
+            block = fluid.default_main_program().global_block()
+
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/op_test.py:1121]: capsulate these code as a method

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 7]
```diff
@@ -277,45 +278,142 @@ def _calc_dygraph_output(self, place, parallel=False, no_check_set=None):
...
             inputs = defaultdict(list)
-            for name, np_value in six.iteritems(self.inputs):
-                if not isinstance(np_value, list):
-                    np_value = [np_value]
-
-                for i in range(len(np_value)):
-                    inputs[name].append(
-                        self._create_var_from_numpy(np_value[i]))
-
+            op_proto = OpProtoHolder.instance().get_op_proto(self.op_type)
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/op_test.py:346]: seal these code as a `method`

**Review by JiabinYang — COMMENTED** [review]
some comment

File: paddle/fluid/operators/crf_decoding_op.h (hunk 1 above)
Comment (JepsonWong) [reply paddle/fluid/operators/crf_decoding_op.h]: done

File: paddle/fluid/operators/crop_op.h (hunk 2 above)
Comment (JepsonWong) [reply paddle/fluid/operators/crop_op.h]: done

File: paddle/fluid/operators/crop_tensor_op.h (hunk 3 above)
Comment (JepsonWong) [reply paddle/fluid/operators/crop_tensor_op.h]: done

File: paddle/fluid/operators/data_norm_op.cc (hunk 4 above)
Comment (JepsonWong) [reply paddle/fluid/operators/data_norm_op.cc]: done

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 5 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/op_test.py]: done

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 6 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/op_test.py:1121]: done

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 7 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/op_test.py:346]: done

**Review by JiabinYang — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/expand_op.h [hunk 8]
```diff
@@ -101,8 +101,9 @@ class ExpandKernel : public framework::OpKernel<T> {
     switch (rank) {
       REP_EXPAND_TEMPLATE(MAX_RANK_SUPPORTED)
       default:
-        PADDLE_ENFORCE(false,
-                       "Only support tensor with rank being between 1 and 6.");
+        PADDLE_ENFORCE_EQ(
+            false, true,
```
Comment (phlrain) [inline paddle/fluid/operators/expand_op.h]: not enfore false true

File: paddle/fluid/operators/expand_op.h [hunk 9]
```diff
@@ -181,8 +182,9 @@ class ExpandGradKernel : public framework::OpKernel<T> {
       switch (dims) {
         REP_EXPAND_GRAD_TEMPLATE(72)
         default:
-          PADDLE_ENFORCE(
-              false, "Only support tensor with rank being between 1 and 6.");
+          PADDLE_ENFORCE_EQ(
```
Comment (phlrain) [inline paddle/fluid/operators/expand_op.h]: not enfore true false

File: paddle/fluid/operators/nce_op.cc [hunk 10]
```diff
@@ -212,6 +213,33 @@ By default this operator uses a uniform distribution for sampling.
...
+    op->SetInput("CustomDistAliasProbs", this->Input("CustomDistAliasProbs"));
+    op->SetInput(framework::GradVarName("Cost"), this->OutputGrad("Cost"));
+    op->SetOutput(framework::GradVarName("Input"), this->InputGrad("Input"));
+    op->SetOutput(framework::GradVarName("Bias"), this->InputGrad("Bias"));
+    op->SetOutput(framework::GradVarName("Weight"), this->InputGrad("Weight"));
+    op->SetAttrMap(this->Attrs());
+    return std::unique_ptr<T>(op);
+  }
+};
+
```
Comment (phlrain) [inline paddle/fluid/operators/nce_op.cc:242]: 这个 和 default grad maker有啥区别

File: paddle/fluid/operators/unpool_op.cc [hunk 11]
```diff
@@ -114,6 +115,23 @@ class UnpoolOp : public framework::OperatorWithKernel {
...
+  using framework::SingleGradOpMaker<T>::SingleGradOpMaker;
+  std::unique_ptr<T> Apply() const override {
+    auto* op = new T();
+    op->SetType(this->ForwardOpType() + "_grad");
+    op->SetInput("X", this->Input("X"));
+    op->SetInput("Indices", this->Input("Indices"));
+    op->SetInput("Out", this->Output("Out"));
+    op->SetInput(framework::GradVarName("Out"), this->OutputGrad("Out"));
+    op->SetOutput(framework::GradVarName("X"), this->InputGrad("X"));
+    op->SetAttrMap(this->Attrs());
```
Comment (phlrain) [inline paddle/fluid/operators/unpool_op.cc:134]: 是不是check下，这种设置 和 default grad maker有啥区别

File: python/paddle/fluid/tests/unittests/mkldnn/test_concat_mkldnn_op.py [hunk 12]
```diff
@@ -24,6 +24,10 @@ def setUp(self):
         self.attrs["use_mkldnn"] = True
         self._cpu_only = True
 
+    def test_check_output(self):
```
Comment (phlrain) [inline python/paddle/fluid/tests/unittests/mkldnn/test_concat_mkldnn_op.py:27]: 这个pass 是不是静态图也不跑了

File: python/paddle/fluid/tests/unittests/mkldnn/test_concat_mkldnn_op.py [hunk 13]
```diff
@@ -50,6 +58,10 @@ def setUp(self):
...
+        # TODO(wangzhongpu): support mkldnn op in dygraph mode 
+        pass
+
     def test_check_grad(self):
```
Comment (phlrain) [inline python/paddle/fluid/tests/unittests/mkldnn/test_concat_mkldnn_op.py:65]: 这个pass 是这个op 没有反向么

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 14]
```diff
@@ -753,7 +873,8 @@ def find_actual(target_name, fetch_list):
                 for item in sub_out:
                     sub_out_name, expect = item[0], item[1]
                     if check_dygraph:
-                        imperative_actual = dygraph_outs[sub_out_name][0]
+                        imperative_actual = find_imperative_actual(
+                            sub_out_name, dygraph_outs, place)
                         imperative_actual_t = np.array(
```
Comment (phlrain) [inline python/paddle/fluid/tests/unittests/op_test.py:831]: 这个地方可以直接 .numpy() 吧

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 15]
```diff
@@ -780,15 +901,17 @@ def find_actual(target_name, fetch_list):
...
                             "Output (" + sub_out_name +
                             ") has different lod at " + str(place))
-                    if check_dygraph:
-                        self.assertListEqual(
-                            imperative_actual._ivar.value().get_tensor()
-                            .recursive_sequence_lengths(), expect[1],
-                            "Output (" + out_name + ") has different lod at " +
-                            str(place) + " in dygraph mode")
+                        if check_dygraph:
+                            self.assertListEqual(
```
Comment (phlrain) [inline python/paddle/fluid/tests/unittests/op_test.py:858]: 这个地方 是判断lod 是否一致么。 lod的结果什么时候存储到 output的 tensor里面了

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 16]
```diff
@@ -803,16 +926,22 @@ def find_actual(target_name, fetch_list):
...
-                            expect_t,
-                            atol=atol,
-                            equal_nan=equal_nan),
-                        "Output (" + out_name + ") has diff at " + str(place) +
-                        "\nExpect " + str(expect_t) + "\n" + "But Got" +
-                        str(imperative_actual_t) + " in class " +
-                        self.__class__.__name__)
+                    if six.moves.reduce(
+                            lambda x, y: x * y, imperative_actual_t.shape,
+                            1) == 0 and six.moves.reduce(
```
Comment (phlrain) [inline python/paddle/fluid/tests/unittests/op_test.py:884]: 这个地方为啥要pass

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 17]
```diff
@@ -952,11 +1084,119 @@ def check_grad_with_place(self,
...
+                block.append_op(
+                    type='scale',
+                    inputs={"X": loss_sum},
+                    outputs={"Out": loss},
+                    attrs={'scale': 1.0 / float(len(avg_sum))})
+            loss.backward()
+
+            fetch_list_grad = []
+            for inputs_to_check_name in inputs_to_check:
+                a = (np.array(inputs_grad_dict[inputs_to_check_name]
```
Comment (phlrain) [inline python/paddle/fluid/tests/unittests/op_test.py]: 拿到grad 之后，直接.numpy() 转换为 numpy值

File: python/paddle/fluid/tests/unittests/test_elementwise_mul_op.py [hunk 18]
```diff
@@ -84,6 +98,7 @@ def setUp(self):
             'Y': np.random.random((32, )).astype("float64")
         }
         self.outputs = {'Out': np.multiply(self.inputs['X'], self.inputs['Y'])}
+        self.init_kernel_type()
```
Comment (phlrain) [inline python/paddle/fluid/tests/unittests/test_elementwise_mul_op.py:101]: 为啥最后调用这个 init_kernel_type

File: paddle/fluid/operators/expand_op.h (hunk 8 above)
Comment (JepsonWong) [reply paddle/fluid/operators/expand_op.h]: done

File: paddle/fluid/operators/expand_op.h (hunk 9 above)
Comment (JepsonWong) [reply paddle/fluid/operators/expand_op.h]: done

File: paddle/fluid/operators/nce_op.cc (hunk 10 above)
Comment (JepsonWong) [reply paddle/fluid/operators/nce_op.cc:242]: 默认grad maker会为所有的Variable构造GRAD Variable，但有些GRAD在 Variable计算过程中没有被用到，所以没有被初始化，而动态图会为所有的GRAD Variable做累加操作；所以在动态图下会出错。

File: paddle/fluid/operators/unpool_op.cc (hunk 11 above)
Comment (JepsonWong) [reply paddle/fluid/operators/unpool_op.cc:134]: 默认grad maker会为所有的Variable构造GRAD Variable，但有些GRAD在 Variable计算过程中没有被用到，所以没有被初始化，而动态图会为所有的GRAD Variable做累加操作；所以在动态图下会出错。

File: python/paddle/fluid/tests/unittests/mkldnn/test_concat_mkldnn_op.py (hunk 12 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/mkldnn/test_concat_mkldnn_op.py:27]: 是的，已修改。

File: python/paddle/fluid/tests/unittests/mkldnn/test_concat_mkldnn_op.py (hunk 13 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/mkldnn/test_concat_mkldnn_op.py:65]: 有反向，但原先这部分没有测试，故沿用之前写法。

File: python/paddle/fluid/tests/unittests/test_elementwise_mul_op.py (hunk 18 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/test_elementwise_mul_op.py:101]: 为了初始化self.use_mkldnn变量。

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 16 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/op_test.py:884]: 这种情况下维度为0，故可以pass掉。
原因在于静态图通过fetch_op拿输出变量，会对维度为0的情况进行处理，而动态图不会为这种情况进行处理，导致动静态图出现偏差。而本身维度为0不需要按位比较，故pass掉。

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 15 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/op_test.py:858]: 在op infershape推导、op计算过程中会写入LoDTensor中。

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 17 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/op_test.py]: done

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 14 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/op_test.py:831]: done

Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/op_test.py]: 测试不可以直接numpy

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 17 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/op_test.py]: 测试不可以直接numpy

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 14 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/op_test.py]: 测试不可以直接numpy

File: paddle/fluid/operators/activation_op.h [hunk 19]
```diff
@@ -60,12 +60,12 @@ inline void ExtractActivationTensor(const framework::ExecutionContext& context,
                                     framework::Tensor** Out) {
   auto x_var = context.InputVar("X");
   auto out_var = context.OutputVar("Out");
-  PADDLE_ENFORCE(x_var != nullptr,
-                 "Cannot get input Variable X, variable name = %s",
-                 context.op().Input("X"));
-  PADDLE_ENFORCE(out_var != nullptr,
-                 "Cannot get output Variable Out, variable name = %s",
-                 context.op().Output("Out"));
+  PADDLE_ENFORCE_EQ(x_var != nullptr, true,
```
Comment (phlrain) [inline paddle/fluid/operators/activation_op.h]: PADDLE_ENFORCE_NE(x_var, nullptr,

File: paddle/fluid/operators/fused/fused_elemwise_activation_op.h [hunk 20]
```diff
@@ -414,13 +415,14 @@ class FusedElemwiseActivationGradKernel : public framework::OpKernel<T> {
  public:
   void Compute(const framework::ExecutionContext &ctx) const override {
     auto in_y = ctx.Input<framework::Tensor>("Y");
-    PADDLE_ENFORCE(in_y != nullptr, "Input(Y) should not be nullptr.");
+    PADDLE_ENFORCE_EQ(in_y != nullptr, true, "Input(Y) should not be nullptr.");
```
Comment (phlrain) [inline paddle/fluid/operators/fused/fused_elemwise_activation_op.h]: PADDLE_ENFORCE_NE

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 21]
```diff
@@ -260,51 +261,115 @@ def _create_var_from_numpy(self, value):
...
+                for (name, np_value) in np_list[name]:
+                    v = create_var(np_value, name, is_input,
+                                   if_return_inputs_grad_dict)
+                    var_list.append(v)
+                    if if_return_inputs_grad_dict:
+                        inputs_grad_dict[name] = v
+                var_dict[slot_name] = var_list
+            else:
+                np_value = np_list[name]
+                if isinstance(np_list[name], list):
```
Comment (phlrain) [inline python/paddle/fluid/tests/unittests/op_test.py:325]: 非 duplicate的情况下 是不是没有必要用for循环了

**Review by phlrain — DISMISSED** [review]

**Review by raindrops2sea — DISMISSED** [review]

File: paddle/fluid/operators/fused/fused_elemwise_activation_op.h (hunk 20 above)
Comment (JepsonWong) [reply paddle/fluid/operators/fused/fused_elemwise_activation_op.h]: done

File: python/paddle/fluid/tests/unittests/op_test.py (hunk 21 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/op_test.py:325]: done

File: paddle/fluid/operators/activation_op.h (hunk 19 above)
Comment (JepsonWong) [reply paddle/fluid/operators/activation_op.h]: done

**Review by phlrain — APPROVED** [review]

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**